psi4/driver/p4util/solvers.py

```diff
@@ -606,48 +623,63 @@ def vector_copy(X):
         """
         pass
 
+    @abstractmethod
+    def residue(self, X, so_prop_ints):
+        """Compute residue
+
+        Parameters
+        ----------
+        X : single `vector`
+          The `vector` to use to compute the property.
+        so_prop_ints :
+          Property integrals in SO basis for the desired transition property.
+        prefactor : float, optional
+          Scaling factor.
+
+        Returns
+        -------
+        residue :
+          The transition property.
+        """
+        pass
 
-def davidson_solver(engine,
-                    guess,
-                    e_tol=1.0E-6,
-                    r_tol=1.0E-8,
-                    nroot=1,
-                    max_vecs_per_root=20,
-                    maxiter=100,
-                    verbose=1,
-                    schmidt_tol=1.0e-8):
-    """
 
-    Solves for the lowest few eigenvalues and eigenvectors of a large problem emulated through an engine.
+def davidson_solver(engine, guess, *, nroot, r_tol=1.0E-4, max_ss_size=100, maxiter=60, verbose=1):
+    """Solves for the lowest few eigenvalues and eigenvectors of a large problem emulated through an engine.
 
 
-    If the large matrix `A` has dimension `{NxN}` and N is very large, and only a small number of roots, `k`
-    are desired this algorithm is preferable to standard methods as uses on the order of `N * k` memory.
-    One only needs to have the ability to compute the product of a times a vector.
+    If the large matrix `A` has dimension `{NxN}` and N is very large, and only
+    a small number of roots, `k` are desired this algorithm is preferable to
+    standard methods as uses on the order of `N * k` memory. One only needs to
+    have the ability to compute the product of a times a vector.
 
-    For non-hermitan `A` the basis of the algorithm breaks down. However in practice, for strongly diagonally-dominant `A`
-    such as the similarity transformed hamiltonian in EOM-CC this algorithm commonly still used.
+    For non-hermitan `A` the basis of the algorithm breaks down. However in
+    practice, for strongly diagonally-dominant `A` such as the
+    similarity-transformed Hamiltonian in EOM-CC this algorithm is commonly still
+    used.
 
     Parameters
     -----------
     engine : object (subclass of :class:`SolverEngine`)
-       The engine drive all operations involving data structures that have at least one "large" dimension. See :class:`SolverEngine` for requirements
+       The engine drive all operations involving data structures that have at
+       least one "large" dimension. See :class:`SolverEngine` for requirements
     guess : list {engine dependent}
        At least `nroot` initial expansion vectors
-    e_tol : float
-        Convergence tolerance for eigenvalues
-    r_tol : float
-        Convergence tolerance for residual vectors
     nroot : int
         Number of roots desired
+    r_tol : float
```

Comment:
* Is this close enough to the general option `r_convergence` that we should call it that (maybe with a TDSCF prefix) for the user?
* If so, do you think it's worth doing a wholesale `r_tol` --> `r_convergence` in the solver code? Or maybe `r_conv` as compromise?
* If r_tol is not touched by user, should its default be (in exponents) SCF `e_convergence` - 2 (or `d_convergence`)?

Reply:
* It is similar to the `r_convergence` option in CC, so I agree to rename to `r_convergence` in all TDDFT-related functions.
* User-facing, I agree on `TDSCF_R_CONVERGENCE`. But maybe this is something for #1885 
* `r_convergence` can be set by the user, but we can come up with a pretty safe default: `min(1e-4, d_convergence*1e2)` this way the user has to explicitly tell Psi4 to seek an insanely tight residual convergence on tight `d_convergence`. I am unsure how to achieve this though. Is there a way I can probe defaults from a PsiAPI function?

Reply:
is something like the below what you're after? upper path will only trigger if user has touched the specific tdscf option. lower path d_conv value already reflects general user setting or the driver adjustment for derivative type a la http://psicode.org/psi4manual/master/scf.html#convergence-and-algorithm-defaults (essentially 1.e-6).
```
if core.has_option_changed('SCF', 'TDSCF_R_CONVERGENCE'):
    r_convergence = core.get_option('SCF', 'TDSCF_R_CONVERGENCE')
else:
    r_convergence = min(1.e-4, core.get_option('SCF', 'D_CONVERGENCE') * 1.e2)
```

Reply:
I think that's what I am after.